### ISSUE TITLE
Batch size error when using GRPOTrainer with vllm

### ISSUE THREAD

**Issue by PerceptionDkh** [huggingface/trl#3181]
### Reproduction

```python
training_args = GRPOConfig(
      num_generations=8,
      per_device_train_batch_size=8,
      use_vllm=True,
      vllm_device="cuda:1",
      ...
    )
trainer = GRPOTrainer(
      model=model,
      reward_funcs=reward_func,
      args=training_args,
      train_dataset=train_dataset,
      eval_dataset=eval_dataset,
      peft_config=peft_config
)
```

When I check the `reward_func` that I've customized, I realized that I received 16 completions, which is not as I wanted (8). Besides, there are 2 types of prompts within the 16 prompts. And when calculating the mean and std of the advantage, all the 16 completions will be used, which is supposed to be 8, because they should be generated according to the same prompt.
I found that in class `TrainingArgument`, `train_batch_size` will be calculated as `per_device_batch_size`*`n_gpu`, which is 8*2=16 in my case. And when generating completions with `vllm`, since I set `vllm_device` to "cuda:1", the large language model will receive 16 prompts (2 types) and generate 16 completions. The 16 completions will all be sent to `reward_func` to calculate rewards. After that, advantages will be calculated and all the 16 completions will be used, which is not right. I think advantage should be calculated from completions related to the same prompt.

### System Info

- Platform: Linux-6.8.0-52-generic-x86_64-with-glibc2.35
- Python version: 3.9.21
- PyTorch version: 2.6.0
- CUDA device(s): NVIDIA A100 80GB PCIe, NVIDIA A100 80GB PCIe
- Transformers version: 4.49.0
- Accelerate version: 1.5.2
- Accelerate config: not found
- Datasets version: 3.4.0
- HF Hub version: 0.29.3
- TRL version: 0.15.2
- bitsandbytes version: 0.45.3
- DeepSpeed version: 0.16.4
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.68.1
- PEFT version: 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete
